src/main/java/com/aventstack/extentreports/concurrent/ReadWriteMap.java

```diff
@@ -0,0 +1,70 @@
+package com.aventstack.extentreports.concurrent;
+
+
+import java.util.Map;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.BiConsumer;
+
+/**
+ * ReadWriteList.java
+ * This class demonstrates how to use ReadWriteLock to add concurrency
+ * features to a non-threadsafe collection
+ *
+ * @author www.codejava.net
+ */
+public class ReadWriteMap<K, V> {
```

Comment:
I was going over the C# code and I observed locks in several places, see here for example:

https://github.com/extent-framework/extentreports-csharp/blob/master/ExtentReports/Core/ExtentObservable.cs#L126

I guess it is to lock the given block to a particular thread and let other threads wait to acquire access.

Reply:
I understand the usage of locks. However, locks are costly. Also, locks can be avoided in some cases with a proper design. There should be a real need to have locks. I am not at that stage in the review process that I can comment of authenticity of the usage of locks. My general idea would be to avoid it unless absolutely necessary.

Reply:
Since my comment was in context of C#, would consider that as a grain of salt as I was told this is done intentionally to be supported by .NET core without any effort.  Note: you cannot synchronize methods in .NET core like in .NET framework so locking is the only option.

For Java, I do not see any locks used but only synchronized methods.  But even in this case with locks, its a small price to pay for thread clashes as mentioned by OP.  I, on the other hand prefer synchronized as its cleaner.  I guess @ anshooarora also prefers it so, as that's what I see from him most of time.

Reply:
But in this case, I feel performance impact would be minimal to none as the locking scope is tiny. 

PS.  Learned something new today, a good read here:  https://plumbr.io/blog/locked-threads/improving-lock-performance-in-java

Reply:
It is not about locking scope, it is about how many times you take the lock.